pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lazyrestmapper
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// LazyRESTMapper is a RESTMapper that will lazily query the provided
+// client for discovery information to do REST mappings.
+type LazyRESTMapper struct {
```

Comment:
Curious if we though actually about making this RESTMapper the default implementation, truthfully it sounds like that the benefit of making less calls, and only making them when needed are something that a lot of users might want. 

If we decide to go down this path, we should however provide a feature gate of sorts that enables this as an experiment first (opt in) -> on by default -> then enable it and remove the gate.

Thoughts?

Reply:
Hey! Initially I didn't plan to make it default but rather to use it either for clusters with a slow API server, or for very crd-pumped ones (ahem, OpenShift), when we have to do more than a hundred requests to initialize the mapper.

But after multiple optimizations I could say there is no worsening relative to the current implementation, only improvements. So, it's reasonable to make it default after some testing.

Do you have an example of such feature gating in this project, so I can mark this feature as experimental?

Reply:
btw, we test this implementation with OpenShift here: https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/213

Reply:
We'd need to introduce a feature gate, which we currently don't have support for. For this PR, we can focus immediately to do some sort of split-logic between how things were done in the default RESTMapper, and this way.

If we feel comfortable though and have enough testing coverage all around, we can mark this a breaking change (⚠️) and add extensive release notes and ask for feedback.

@alvaroaleman wdyt?

Reply:
I'd suggest to simply add this restmapper for now and backport it into 0.14. 0.15 onwards (or later, depending on if we find issues) we can simply make it the default but leave the option to provide a custom one.

Reply:
We can definitely backport, although for 0.15 (next version) I'd prefer to have a single RESTMapper, and merge the current Dynamic with the Lazy implementation. We've only ever exposed a single one, and I worry that more than one RESTMapper implementation might create even more confusion than we have today. 

The other option I see is to just merge this as net-new addition to the DynamicRESTMapper implementation, make it default in 0.15 and wait for feedback. 

The current DynamicRESTMapper implementation godoc reads:
```
dynamicRESTMapper is a RESTMapper that dynamically discovers resource types at runtime.
```
which makes it sound like it's dynamic and lazy, an on-demand type rest mapper, but it's really not.

Reply:
Merging the two for next release sounds good to me.

I _think_ the new lazy restmapper does everything the current dynamic one does btw. The current dynamic one will on a resource not found error reload the entire restmapping (with some ratelimiting). Since the new one will simply try to load the restmapping on-demand if it doesn't have it cached, that should solve the same problem just as well.

Reply:
Could we have the implementation of this PR inlined with the current one, but all of it behind a feature gate?

Reply:
How would you featuregate this?

Reply:
A new `DynamicRESTMapperOption` that sets the flag could work?

Reply:
Something like: WithExperimentalLazyMapper

Reply:
sgtm